### PR TITLE
Only store tracking spans in the editor side of nav bar items

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This is part of a larger refactoring for navbars.  The final form will be that a navbar item can store:

1. teh full (tracking) span(s) in this document that will be used to determine if the item should be selected based on user position.
2. the (tracking) span in the document to navigate to if the item is selected.
3. the doc-id + span if the item should cause navigation to another file (for example, in partial cases).

this PR does '1'.

### CODE REVIEW

File: src/EditorFeatures/Core/Extensibility/NavigationBar/AbstractEditorNavigationBarItemService.cs [hunk 1]
```diff
@@ -26,15 +28,16 @@ protected AbstractEditorNavigationBarItemService(IThreadingContext threadingCont
         protected abstract Task<VirtualTreePoint?> GetSymbolNavigationPointAsync(Document document, ISymbol symbol, CancellationToken cancellationToken);
         protected abstract Task<bool> TryNavigateToItemAsync(Document document, WrappedNavigationBarItem item, ITextView textView, CancellationToken cancellationToken);
 
-        public async Task<IList<NavigationBarItem>?> GetItemsAsync(Document document, CancellationToken cancellationToken)
+        public async Task<ImmutableArray<NavigationBarItem>> GetItemsAsync(
+            Document document, ITextSnapshot textSnapshot, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Extensibility/NavigationBar/AbstractEditorNavigationBarItemService.cs:32]: note: this is the *editor* portion of the API.  it does pass along the textsnaphsot now as the *Editor* portion returns navbaritems that have tracking spans in them.  so this allows us to create those tracking spans at construction time, not some point in the future after.  the *feature* api is of course oblivious to this.

File: src/EditorFeatures/Core/Extensibility/NavigationBar/AbstractEditorNavigationBarItemService.cs [hunk 2]
```diff
@@ -26,15 +28,16 @@ protected AbstractEditorNavigationBarItemService(IThreadingContext threadingCont
...
         {
             var service = document.GetRequiredLanguageService<CodeAnalysis.NavigationBar.INavigationBarItemService>();
             var workspaceSupportsDocumentChanges = document.Project.Solution.Workspace.CanApplyChange(ApplyChangesKind.ChangeDocument);
             var items = await service.GetItemsAsync(document, workspaceSupportsDocumentChanges, cancellationToken).ConfigureAwait(false);
-            return items.SelectAsArray(v => (NavigationBarItem)new WrappedNavigationBarItem(v));
+            return items.SelectAsArray(v => (NavigationBarItem)new WrappedNavigationBarItem(v, textSnapshot));
         }
 
-        public Task<bool> TryNavigateToItemAsync(Document document, NavigationBarItem item, ITextView textView, CancellationToken cancellationToken)
+        public Task<bool> TryNavigateToItemAsync(Document document, NavigationBarItem item, ITextView textView, ITextSnapshot textSnapshot, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Extensibility/NavigationBar/AbstractEditorNavigationBarItemService.cs:40]: navigation also takes in the text snapshot so we can determine teh locatin of an item at the point that we navigate.  However, currently only TS and F# use this.  Roslyn always goes back to symbols to navigate.  that will be fixed in one of the followup prs coming shortly.

File: src/EditorFeatures/Core/Extensibility/NavigationBar/INavigationBarItemService.cs [hunk 3]
```diff
@@ -2,22 +2,23 @@
...
     {
-        Task<IList<NavigationBarItem>?> GetItemsAsync(Document document, CancellationToken cancellationToken);
+        Task<ImmutableArray<NavigationBarItem>> GetItemsAsync(Document document, ITextSnapshot textSnapshot, CancellationToken cancellationToken);
         bool ShowItemGrayedIfNear(NavigationBarItem item);
 
         /// <summary>
         /// Returns <see langword="true"/> if navigation (or generation) happened.  <see langword="false"/> otherwise.
         /// </summary>
-        Task<bool> TryNavigateToItemAsync(Document document, NavigationBarItem item, ITextView view, CancellationToken cancellationToken);
+        Task<bool> TryNavigateToItemAsync(Document document, NavigationBarItem item, ITextView view, ITextSnapshot textSnapshot, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Extensibility/NavigationBar/INavigationBarItemService.cs:22]: these apis are EA, so it's fine to change them.

File: src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItem.cs [hunk 4]
```diff
@@ -21,38 +18,27 @@ internal abstract class NavigationBarItem
         public int Indent { get; }
         public ImmutableArray<NavigationBarItem> ChildItems { get; }
 
-        public ImmutableArray<TextSpan> Spans { get; internal set; }
-        internal ImmutableArray<ITrackingSpan> TrackingSpans { get; set; } = ImmutableArray<ITrackingSpan>.Empty;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItem.cs:25]: no need for 'Spans' anymore in the editor api.  nothing uses that.  ANd TRackingSpans is no longer mutable as it is provided at construction time.

File: src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs [hunk 5]
```diff
@@ -383,7 +377,6 @@ private async Task ProcessItemSelectionAsync(NavigationBarItem item, Cancellatio
                 {
                     var navBarService = document.GetRequiredLanguageService<INavigationBarItemService>();
                     var snapshot = _subjectBuffer.CurrentSnapshot;
-                    item.Spans = item.TrackingSpans.SelectAsArray(ts => ts.GetSpan(snapshot).Span.ToTextSpan());
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/NavigationBar/NavigationBarController.cs:386]: how we hackily used to do things.

**Review by dibarbet — APPROVED** [review]

File: src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItem.cs [hunk 6]
```diff
@@ -21,38 +18,27 @@ internal abstract class NavigationBarItem
...
-
-        // Legacy constructor for TypeScript.
-        [Obsolete("This is a compatibility shim for TypeScript; please do not use it.")]
-        public NavigationBarItem(string text, Glyph glyph, IList<TextSpan> spans, IList<NavigationBarItem>? childItems = null, int indent = 0, bool bolded = false, bool grayed = false)
-            : this(text, glyph, spans.ToImmutableArrayOrEmpty(), childItems.ToImmutableArrayOrEmpty(), indent, bolded, grayed)
-        {
-        }
+        public ImmutableArray<ITrackingSpan> TrackingSpans { get; }
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/Core/Extensibility/NavigationBar/NavigationBarItem.cs:21]: Yup.  It' in the next PR :)

**Pull request merged**